design/ironic-standalone-operator.md

```diff
@@ -0,0 +1,369 @@
+each Ironic instance must register its RPC access IP or hostname in the
+database.
+
+When TLS support is enabled, the RPC communication must be secured by
+TLS as well. This may pose a problem since each Ironic instance needs a TLS
+certificate that is valid for its RPC access IP or hostname. This problem has
+been extensively discussed in the [initial HA proposal][issue3], and here is
+the proposed solution, at least for the MVP case:
+
+The Ironic controller will generate a self-signed CA and pass its public and
+private parts into each Ironic container. Each Ironic container will generate
+its private key certificate and sign the certificate with this CA. The CA will
+be trusted **only** for the RPC purpose, removing the possibility of abuse.
+To reduce the number of code paths, this process will happen unconditionally,
+even when TLS for Ironic itself is not enabled.
+
+[kea]: https://www.isc.org/kea/
+[issue3]: https://github.com/metal3-io/ironic-standalone-operator/issues/3
+
+#### Architecture FAQ
+
+Q: Why cannot we split dnsmasq into a separate deployment in the non-HA
+architecture? A: That may require having more than one IP address on the
+provisioning network: for dnsmasq and for httpd/ironic. This is a new
+operational requirement that I'd like to avoid at this stage.
+
+Q: Why does the same dnsmasq limitation affect the HA architecture. A: The HA
+architecture is completely new here, so we can introduce new requirements
+without regression in the operational experience.
+
+Q: Why using *DaemonSets* if *StatefulSets* provide us an easier way to address
+separate Ironic instances? A: While we're relying on host networking, making
+several Ironic instances co-exist on the same Kubernetes node is too complex.
+
+Q: Why aren't we using HostPort services? A: The fact that they provide a
+random port is a roadblock for production deployments since many of them
+require opening a predictable port in the firewall configuration. If we use a
+pre-defined port, it may cause conflicts with other HostPort services or even
+end up outside of the allowed range.
+
+#### Current API design
+
+Currently, the API consists of two main objects: `IronicDatabase` and `Ironic`.
+
+The `IronicDatabase` object is very simple:
+
+- `credentialsRef` - a reference to a secret with credentials (generated if
+  missing)
+- `image` - container image to use
+- `tlsRef` - a reference to a TLS secret to use for the service
+
+The `Ironic` object is much more complex and should probably be split into more
+custom resources as we polish its internal architecture. Currently, it uses
+nested structures to logically group fields. Here are the most important fields
+(omitting various fine-tuning for brevity):
+
+- `credentialsRef` - a reference to a secret with credentials (generated if
+  missing)
+- `databaseRef` - a reference to an `IronicDatabase` object (if needed)
+- `distributed` - a boolean flag that enables the HA architecture
```

Comment:
Why not just call this field `highlyAvailable` or `highAvailability`

Reply:
That's a good question. I guess `distributed` is more precise technically, while `highAvailability` may be more user-friendly. I'd like to hear more opinions on the topic.

Reply:
I would vote for highAvailability or haEnabled

Reply:
Okay, let's go with highAvailability.